src/models/StyledComponent.js

```diff
@@ -150,37 +192,32 @@ export default (ComponentStyle: Function, constructWithOptions: Function) => {
 
           this.setState({ theme, generatedClassName })
         })
-      } else {
-        // eslint-disable-next-line react/prop-types
-        const theme = this.props.theme || {}
-        const generatedClassName = this.generateAndInjectStyles(
-          theme,
-          this.props
-        )
-        this.setState({ theme, generatedClassName })
       }
     }
 
-    componentWillReceiveProps(nextProps: {
-      theme?: Theme,
-      [key: string]: any,
-    }) {
+    componentDidUpdate(prevProps: { theme?: Theme, [key: string]: any }) {
       // If this is a staticaly-styled component, we don't need to listen to
       // props changes to update styles
       const { componentStyle } = this.constructor
       if (componentStyle.isStatic) {
         return
       }
 
+      // Only react to props change to avoid infinite update loop
+      if (prevProps === this.props) {
+        return
+      }
+
+      // eslint-disable-next-line react/no-did-update-set-state
       this.setState(oldState => {
         const theme = determineTheme(
```

Comment:
setState in cDU typically leads to re render. Why isn’t this using gDSFP instead?

Reply:
Just forgot to remove it in https://github.com/styled-components/styled-components/pull/1664/commits/ed7efd2c019eba0817d42bd6bf82c7ba0543f704, will do